cache/blobs.go

```diff
@@ -59,6 +60,177 @@ func (sr *immutableRef) computeBlobChain(ctx context.Context, createIfNeeded boo
+			return zstdWriter(comp)(dest)
+		}
+		mediaType = ocispecs.MediaTypeImageLayer + "+zstd"
+	case compression.Nydus:
+		compressorFunc, finalize = compressNydus(ctx, comp)
+		mediaType = nydusify.MediaTypeNydusBlob
+	default:
+		return nil, errors.Errorf("unknown layer compression type: %q", comp.Type)
+	}
+
+	var lowerRef *immutableRef
+	switch sr.kind() {
+	case Diff:
+		lowerRef = sr.diffParents.lower
+	case Layer:
+		lowerRef = sr.layerParent
+	}
+	var lower []mount.Mount
+	if lowerRef != nil {
+		m, err := lowerRef.Mount(ctx, true, s)
+		if err != nil {
+			return nil, err
+		}
+		var release func() error
+		lower, release, err = m.Mount()
+		if err != nil {
+			return nil, err
+		}
+		if release != nil {
+			defer release()
+		}
+	}
+
+	var upperRef *immutableRef
+	switch sr.kind() {
+	case Diff:
+		upperRef = sr.diffParents.upper
+	default:
+		upperRef = sr
+	}
+	var upper []mount.Mount
+	if upperRef != nil {
+		m, err := upperRef.Mount(ctx, true, s)
+		if err != nil {
+			return nil, err
+		}
+		var release func() error
+		upper, release, err = m.Mount()
+		if err != nil {
+			return nil, err
+		}
+		if release != nil {
+			defer release()
+		}
+	}
+
+	var desc ocispecs.Descriptor
+	var err error
+
+	// Determine differ and error/log handling according to the platform, envvar and the snapshotter.
+	var enableOverlay, fallback, logWarnOnErr bool
+	if forceOvlStr := os.Getenv("BUILDKIT_DEBUG_FORCE_OVERLAY_DIFF"); forceOvlStr != "" && sr.kind() != Diff {
+		enableOverlay, err = strconv.ParseBool(forceOvlStr)
+		if err != nil {
+			return nil, errors.Wrapf(err, "invalid boolean in BUILDKIT_DEBUG_FORCE_OVERLAY_DIFF")
+		}
+		fallback = false // prohibit fallback on debug
+	} else if !isTypeWindows(sr) {
+		enableOverlay, fallback = true, true
+		switch sr.cm.Snapshotter.Name() {
+		case "overlayfs", "stargz":
+			// overlayfs-based snapshotters should support overlay diff except when running an arbitrary diff
+			// (in which case lower and upper may differ by more than one layer), so print warn log on unexpected
+			// failure.
+			logWarnOnErr = sr.kind() != Diff
+		case "fuse-overlayfs", "native":
+			// not supported with fuse-overlayfs snapshotter which doesn't provide overlayfs mounts.
+			// TODO: add support for fuse-overlayfs
+			enableOverlay = false
+		}
+	}
+	if enableOverlay {
+		computed, ok, err := sr.tryComputeOverlayBlob(ctx, lower, upper, mediaType, sr.ID(), compressorFunc)
+		if !ok || err != nil {
+			if !fallback {
+				if !ok {
+					return nil, errors.Errorf("overlay mounts not detected (lower=%+v,upper=%+v)", lower, upper)
+				}
+				if err != nil {
+					return nil, errors.Wrapf(err, "failed to compute overlay diff")
+				}
+			}
+			if logWarnOnErr {
+				logrus.Warnf("failed to compute blob by overlay differ (ok=%v): %v", ok, err)
+			}
+		}
+		if ok {
+			desc = computed
+		}
+	}
+
+	if desc.Digest == "" && !isTypeWindows(sr) && (comp.Type == compression.Zstd || comp.Type == compression.EStargz) {
```

Comment:
nydus should be listed here as well.
Currently nydus doesn't work on containerd worker + non-overlayfs differ configuration:

```
# mkdir -p /tmp/ctx
# cat <<EOF > /tmp/ctx/Dockerfile
FROM registry2-buildkit:5000/ubuntu:20.04
RUN echo hello > /hello
EOF
# buildctl build --progress=plain --frontend=dockerfile.v0 --local context=/tmp/ctx --local dockerfile=/tmp/ctx \
                  --output type=image,name=registry2-buildkit:5000/ubuntu:20.04-hello-nydus,push=true,compression=nydus,oci-mediatypes=true
#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 103B done
#1 DONE 0.1s

#2 [internal] load .dockerignore
#2 transferring context: 2B done
#2 DONE 0.1s

#3 [internal] load metadata for registry2-buildkit:5000/ubuntu:20.04
#3 DONE 0.1s

#4 [auth] sharing credentials for registry2-buildkit:5000
#4 DONE 0.0s

#5 [1/2] FROM registry2-buildkit:5000/ubuntu:20.04@sha256:adf73ca014822ad8237623d388cedf4d5346aa72c270c5acc01431cc93e18e2d
#5 resolve registry2-buildkit:5000/ubuntu:20.04@sha256:adf73ca014822ad8237623d388cedf4d5346aa72c270c5acc01431cc93e18e2d 0.0s done
#5 sha256:345e3491a907bb7c6f1bdddcf4a94284b8b6ddd77eb7d93f09432b17b20f2bbe 24.12MB / 28.54MB 0.2s
#5 extracting sha256:345e3491a907bb7c6f1bdddcf4a94284b8b6ddd77eb7d93f09432b17b20f2bbe
#5 extracting sha256:345e3491a907bb7c6f1bdddcf4a94284b8b6ddd77eb7d93f09432b17b20f2bbe 0.8s done
#5 extracting sha256:57671312ef6fdbecf340e5fed0fb0863350cd806c92b1fdd7978adbd02afc5c3
#5 extracting sha256:57671312ef6fdbecf340e5fed0fb0863350cd806c92b1fdd7978adbd02afc5c3 0.2s done
#5 extracting sha256:5e9250ddb7d0fa6d13302c7c3e6a0aa40390e42424caed1e5289077ee4054709
#5 extracting sha256:5e9250ddb7d0fa6d13302c7c3e6a0aa40390e42424caed1e5289077ee4054709 0.2s done
#5 DONE 1.5s

#6 [2/2] RUN echo hello > /hello
#6 DONE 0.3s

#7 exporting to image
#7 exporting layers 0.1s done
#7 ERROR: failed to ensure compression type of "nydus": failed to convert: unsupported diff media type: application/vnd.oci.image.layer.nydus.blob.v1: not implemented
------
 > exporting to image:
------
error: failed to solve: failed to ensure compression type of "nydus": failed to convert: unsupported diff media type: application/vnd.oci.image.layer.nydus.blob.v1: not implemented
```

Reply:
Thanks, fixed.